loopy/symbolic.py

```diff
@@ -938,44 +1003,57 @@ class EvaluatorWithDeficientContext(PartialEvaluationMapper):
 
     Returns the expression with the values mapped from :attr:`context`.
     """
-    def map_variable(self, expr):
+
+    @override
+    def map_variable(self, expr: p.Variable) -> Expression:
         if expr.name in self.context:
             return self.context[expr.name]
         else:
             return expr
 
 
 class VariableInAnExpression(CombineMapper[bool, []]):
+    variables_to_search: Collection[p.Variable]
+
     def __init__(self, variables_to_search: Collection[p.Variable]) -> None:
-        assert all(isinstance(variable, p.Variable) for variable in
-            variables_to_search)
+        assert all(isinstance(variable, p.Variable) for variable in variables_to_search)
+
+        super().__init__()
         self.variables_to_search = variables_to_search
 
-    def combine(self, values) -> bool:
+    @override
+    def combine(self, values: Iterable[bool]) -> bool:
         return any(values)
 
-    def map_variable(self, expr) -> bool:
+    @override
+    def map_variable(self, expr: p.Variable) -> bool:
         return expr in self.variables_to_search
 
-    def map_constant(self, expr) -> bool:
+    @override
+    def map_constant(self, expr: object) -> bool:
         return False
 
 
 class SweptInameStrideCollector(CoefficientCollectorBase):
     """
     Mapper to compute the coefficient swept inames for :class:`SubArrayRef`.
     """
-    def map_algebraic_leaf(self, expr):
+
+    @override
+    def map_algebraic_leaf(self, expr: p.AlgebraicLeaf) -> CoeffsT:
         # subscripts that are not involved in :attr:`target_names` are treated
         # as constants.
-        if isinstance(expr, p.Subscript) and (self.target_names is None
-                or expr.aggregate.name not in self.target_names):
+        if isinstance(expr, p.Subscript) and (
+                self.target_names is None
+                or getattr(expr.aggregate, "name", None) not in self.target_names):
             return {1: expr}
 
         return super().map_algebraic_leaf(expr)
 
 
-def get_start_subscript_from_sar(sar, kernel):
+def get_start_subscript_from_sar(
+            sar: SubArrayRef, kernel: LoopKernel,
+        ) -> p.Variable | p.Subscript:
     """
     Returns an instance of :class:`pymbolic.primitives.Subscript`, the
     beginning subscript of the array swept by the *SubArrayRef*.
```

Comment:
Is this correct? The docs seem to say that only `Subscript` can be returned, but the tests failed with the assert below.

Reply:
>  the tests

Which tests?

> failed with the assert below.

What assert?

It really ought to return a `Subscript`. (I.e., not even a `Variable`.) I can't think of a case where it would not?

Reply:
This is is a failed CI that seems to be from the assert I added in this function:
https://github.com/inducer/loopy/actions/runs/18939593172/job/54074769860?pr=966

Not sure how to link to the lines in the PR, but it's just this new assert here:
https://github.com/alexfikl/loopy/blob/07eb2acbeaddc9f21724da74560795e578e8b00f/loopy/symbolic.py#L1073

Reply:
Did some minimal digging in here. It seems like the main issue is that the subscript is empty for the failing test case and `Variable[index]` will just return the variable back. I tried doing a `Variable[EmptyOK(index)]`, but that seems to break a bunch of other stuff.. :\